README.md
# CoreBreakout

[![status](https://joss.theoj.org/papers/add2021f95268fd4cd2850b105f3d570/status.svg)](https://joss.theoj.org/papers/add2021f95268fd4cd2850b105f3d570)

Requirements, installation, and contribution guidelines can be found below. Our full usage and API documentation can be found at: [corebreakout.readthedocs.io](https://corebreakout.readthedocs.io/en/latest/)

### Overview

`corebreakout` is a Python package built around [matterport/Mask\_RCNN](https://github.com/matterport/Mask_RCNN) for the segmentation and depth-alignment of geological core sample images. It provides utilities and an API to enable the workflow depicted in the figure below, as well as a `CoreColumn` data structure to manage and manipulate the resulting depth-registered image data:

![](JOSS_figure_workflow.png)

We are currently using this package to enable research on [Lithology Prediction of Slabbed Core Photos Using Machine Learning Models](https://figshare.com/articles/Lithology_Prediction_of_Slabbed_Core_Photos_Using_Machine_Learning_Models/8023835/2), and are working on getting a DOI for the project through the [Journal of Open Source Software](https://joss.theoj.org/).

## Getting Started

### Target Platform

This package was developed on Linux (Ubuntu, PopOS), and has also been tested on OS X. It may work on other platforms, but we make no guarantees.

### Requirements

In addition to Python`>=3.6`, the packages listed in [requirements.txt](requirements.txt) are required. Notable exceptions to the list are:

- `1.3<=tensorflow-gpu<=1.14` (or possibly just `tensorflow`)
- `mrcnn` via [submodule: matterport/Mask\_RCNN](https://github.com/matterport/Mask_RCNN/tree/3deaec5d902d16e1daf56b62d5971d428dc920bc)

The TensorFlow requirement is not explicitly listed in `requirements.txt` due to the ambiguity between `tensorflow` and `tensorflow-gpu` in versions `<=1.14`. The latter is almost certainly required for training new models, although it may be possible to perform inference with saved models on CPU, and use of the `CoreColumn` data structure does not require a GPU.

Note that TensorFlow GPU capabilities are implemented with [CUDA](https://developer.nvidia.com/cuda-zone), which requires a [supported NVIDIA GPU](https://developer.nvidia.com/cuda-gpus).

#### Additional (Optional) Requirements

Optionally, `jupyter` is required to run demo and test notebooks, and `pytest` is required to run unit tests. Both of these should be manually installed if you plan to modify or contribute to the package source code.

We also provide a script for extraction of top/base depths from core image text using `pytesseract`. After installing the [Tesseract OCR Engine](https://github.com/tesseract-ocr/tesseract) on your machine, you can install the `pytesseract` package with `conda` or `pip`.

### Download code

```
$ git clone --recurse-submodules https://github.com/rgmyr/corebreakout.git
$ cd corebreakout
```

### Download data (optional)

To make use of the provided dataset and model, or to train new a model starting from the pretrained COCO weights, you will need to download the `assets.zip` folder from the [v0.2 Release](https://github.com/rgmyr/corebreakout/releases/tag/v0.2).

Unzip and place this folder in the root directory of the repository (its contents will be ignored by `git` -- see the `.gitignore`). If you would like to place it elsewhere, you should modify the paths in [corebreakout/defaults.py](https://github.com/rgmyr/corebreakout/blob/master/corebreakout/defaults.py) to point to your preferred location.

The current version of `assets/data` has JSON annotation files which include an `imageData` field representing the associated images as strings. For now you can delete this field and reduce the size of the data with `scripts/prune_imageData.py`:

```
$ python scripts/prune_imageData.py assets/
```

### Installation

We recommend installing `corebreakout` and its dependencies in an isolated environment, and further recommend the use of `conda`. See [Conda: Managing environments](https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html).

---

To create a new `conda` environment called `corebreakout-env` and activate it:

```
$ conda create -n corebreakout-env python=3.6 tensorflow-gpu=1.14
$ conda activate corebreakout-env
```

**Note:** If you want to try a CPU-only installation, then replace `tensorflow-gpu` with `tensorflow`. You may also lower the version number if you are on a machine with `CUDA<10.0` (required for TensorFlow`>=1.13`). See [TensorFlow GPU requirements](https://www.tensorflow.org/install/gpu#software_requirements) for more compatibility details.

---

Then install the rest of the required packages into the environment:

```
$ conda install --file requirements.txt
```

---

Finally, install `mrcnn` and `corebreakout` using `pip`. Develop mode installation (`-e`) is recommended (but not required) for `corebreakout`, since many users will want to change some of the default parameters to suit their own data without having to reinstall afterward:

```
$ pip install ./Mask_RCNN
$ pip install -e .
```

## Usage

Please refer to our [readthedocs page](https://corebreakout.readthedocs.io/en/latest/) for full documentation!

## Development and Community Guidelines

### Submit an Issue

- Navigate to the repository's [issue tab](https://github.com/rgmyr/corebreakout/issues)
- Search for existing related issues
- If necessary, create and submit a new issue

### Contributing

- Please see [`CONTRIBUTING.md`](CONTRIBUTING.md) and the [Code of Conduct](CODE_OF_CONDUCT.md) for how to contribute to the project

### Testing

- Most `corebreakout` functionality not requiring trained model weights can be verified with `pytest`:

```
$ cd <root_directory>
$ pytest .
```

- Model usage via the `CoreSegmenter` class can be verified by running `tests/notebooks/test_inference.ipynb` (requires saved model weights)
- Plotting of `CoreColumn`s can be verified by running `tests/notebooks/test_plotting.ipynb`


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## 0.4 [Unreleased]

### To-Do

- Refactor `PolygonDataset` to use `labelme`'s `imageData` field rather than require jpegs

## 0.3

### Changed

- Increased flexibility of included scripts
- Converted `docs/*.md` files to `*.rst`
- Minor paper and bibliography corrections/fixes

### Added

- Built `sphinx`-generated html docs, hosted on `readthedocs`
- Implemented `CoreColumn.iter_chunks()`
- Added demo/inspection notebooks:
  - `column_demo.ipynb`
  - `segmenter_demo.ipynb`
  - `inspect_dataset.ipynb`
- Added explanatory Markdown to `select_model.ipynb`
- Added restriction in `setup.py` to require `tensorflow` before install

## 0.2

### Changed

- Modified code format using `black`
- `README.md` installation instructions
- moved paper files to root directory

### Added

- `requirements.txt` (except `imgaug`, which is included in `setup.py`)
- CONTRIBUTING, LICENSE, and CODE_OF_CONDUCT files
- `tests/notebooks`
- tests for `CoreColumn` saving and loading

### Removed

- superfluous notebooks and scripts

## 0.1

Initial private version.


corebreakout/defaults.py
"""
Default paths, dataset and Mask_RCNN model config, `CoreColumn` viz settings.
"""
import os
from pathlib import Path

from mrcnn.config import Config

from corebreakout import __file__ as PKG_FILE


####+++++++++++++++####
#### Default Paths ####
####+++++++++++++++####

ASSETS_DIR = Path(PKG_FILE).parent.parent / "assets"

# Where to find train/test subdirectories
DATASET_DIR = ASSETS_DIR / "data"

# Where to find saved model weights
MODEL_DIR = ASSETS_DIR / "models"
# Default "core" model weights
CB_MODEL_PATH = MODEL_DIR / 'mask_rcnn_cb_default.h5'
# Pretrained COCO model weights
COCO_MODEL_PATH = MODEL_DIR / 'mask_rcnn_coco.h5'

# Where to save Mask RCNN training checkpoints, etc.
TRAIN_DIR = MODEL_DIR


####++++++++++++++++++++++++####
#### Default Dataset Params ####
####++++++++++++++++++++++++####

CLASSES = ["col", "tray"]

# See `docs/layout_parameters.md` for more information
LAYOUT_PARAMS = {
    "order": "t2b",  # depth order by which to sort set of columns
    "orientation": "l2r",  # depth orientation of each individual column
    "col_height": 1.0,  # assumed height of each column, or tray, etc.
    "col_class": "col",  # name of class for core sample columns
    "endpts": "tray",  # name of class, 'auto', 'auto_all', or 2-tuple
}

####++++++++++++++++++++++####
#### Default Model Config ####
####++++++++++++++++++++++####


class DefaultConfig(Config):
    """M-RCNN model configuration.

    Override some default Mask_RCNN `Config` values.

    For all available parameters and explanations, see:
        https://github.com/matterport/Mask_RCNN/blob/master/mrcnn/config.py
    """

    NAME = "cb_default"

    # Number of classes, including background
    NUM_CLASSES = 1 + len(CLASSES)

    BACKBONE = "resnet101"

    # Length of square anchor side in pixels
    RPN_ANCHOR_SCALES = (64, 128, 192, 256, 320)

    # < 1 : wide anchor, > 1 : tall anchor
    # These defaults assume horizontal (wide) columns
    # Note: starting from COCO model requires exactly 3 anchor ratios
    RPN_ANCHOR_RATIOS = [0.2, 0.5, 1]

    # Non-max suppresion threshold. Increasing generates more proposals.
    RPN_NMS_THRESHOLD = 0.9  # default = 0.7

    # STD_DEVs? Probably not, shouldn't make a big difference.

    # Maximum number of detections and minimum confidence
    DETECTION_MAX_INSTANCES = 6
    DETECTION_MIN_CONFIDENCE = 0.98

    # Set to default number of train/test images, respectively
    # (Can increase former to validate less often though)
    STEPS_PER_EPOCH = 25
    VALIDATION_STEPS = 5

    # Modify loss weights for more precise optimization
    # Few classes present, so we can lower `class` losses
    LOSS_WEIGHTS = {
        "rpn_class_loss": 0.5,
        "rpn_bbox_loss": 1.0,
        "mrcnn_class_loss": 0.5,
        "mrcnn_bbox_loss": 1.0,
        "mrcnn_mask_loss": 1.0,
    }

    # Conservative batch size + assuming single GPU.
    GPU_COUNT = 1
    IMAGES_PER_GPU = 1


####++++++++++++++++++++++++++++++++####
#### Default CoreColumn plot params ####
####++++++++++++++++++++++++++++++++####
"""Set the default parameters for (fig, ax) returned by `CoreColumn.plot()`.

DEPTH_TICK_ARGS -- passed to `viz.make_depth_ticks()`
MAJOR_TICK_PARAMS -- passed to `ax.tick_params(which='major', ...)`
MINOR_TICK_PARAMS -- passed to `ax.tick_params(which='minor', ...)`

You can also add additional arguments to both `*_TICK_PARAMS` from:
    https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.axes.Axes.tick_params.html
"""
DEPTH_TICK_ARGS = {
    "major_precision": 0.1,
    "major_format_str": "{:.1f}",
    "minor_precision": 0.01,
    "minor_format_str": "{:.2f}",
}

MAJOR_TICK_PARAMS = {
    "labelleft": True,  # False to disable labels
    "labelsize": 32,
    "labelcolor": "black",
    "left": True,  # False to disable ticks
    "length": 35,
    "width": 4,
    "color": "black",
}

MINOR_TICK_PARAMS = {
    "labelleft": True,  # False to disable labels
    "labelsize": 12,
    "labelcolor": "black",
    "left": True,  # False to disable ticks
    "length": 5,
    "width": 4,
    "color": "black",
}
